feature_engine/dataframe_checks.py

```diff
@@ -129,3 +129,37 @@ def _check_contains_inf(X: pd.DataFrame, variables: List[Union[str, int]]) -> No
             "Some of the variables to transform contain inf values. Check and "
             "remove those before using this transformer."
         )
+
+
+def _check_X_y_pd_np_mismatch(
+    X: Union[pd.DataFrame, pd.Series, np.ndarray],
+    y: Union[pd.DataFrame, pd.Series, np.ndarray],
+):
+    """
+    Handles case where X is an ndarray and y is a Series,
+    or when X is a DataFrame but y is in an ndarray.
+    In both cases, the non-pandas object will be converted
+    to a pandas object, and take on :wqthe index of the other (pandas) object.
+    If both are ndarray objects, they are returned unchanged.
+    If both are not pandas objects, method returns objects unchanged.
+
+    Parameters
+    ----------
+    X: Pandas DataFrame, Series, or numpy ndarray
+    y: Pandas DataFrame, Series, or numpy ndarray
+
+    Returns
+    -------
+    X: changed as per description above
+    y: changed as per description above
+    """
+    if isinstance(X, np.ndarray) and isinstance(y, (pd.DataFrame, pd.Series)):
+        # already know X is not a DataFrame, use machinery in _is_dataframe()
+        # to correctly convert X to DataFrame
+        X = _is_dataframe(X)
+        X.index = y.index
+    elif isinstance(X, (pd.DataFrame, pd.Series)) and isinstance(y, np.ndarray):
+        y = pd.Series(y)
+        y.index = X.index
+
```

Comment:
I think, given the scenarios that I posed in #376 the best solution would be to replace _is_dataframe in the encoders by a new function that checks simultaneously X and y, because depending on the input combination we should raise errors or not.

thoughts?

Reply:
> X can only be a pandas dataframe at this stage, because if it was an array, is_dataframe converted it to a df
In commits from earlier, I deliberately put the call to `_check_X_y_pd_np_mismatch()`  before the call to `self._check_fit_input_and_variables()` (and thus `_is_dataframe()`) so that `X` has the chance to arrive at `_check_X_y_pd_np_mismatch()` as an array. (For an example, see [here](https://github.com/noahjgreen295/feature_engine/blob/8d6ca178b49684a51bf2eac0c27acffd03195514/feature_engine/encoding/mean_encoding.py#L131) ) This allows me to detect that particular error case.

> and if it was something else that is not permitted, like a pd.Series, is_dataframe should have raised an error (if it does not, we need to fix is_dataframe)
This is correct. I'll remove the logic that handles cases of `X` being a `Series` since that is not possible.

I have additional followup on the [#376](https://github.com/feature-engine/feature_engine/issues/376) page.

Reply:
I would not change is_dataframe because it is used all over our codebase.

I would do the following:

if the encoder needs X and y, instead of using _is_dataframe, use directly your check and return a pandas dataframe and a pd series.

If encoder does not need y: then use is_dataframe as usual.

If encoder has the option to do both (OrdinalEncoder):
```
if self.encoding_method == "ordered":
    X, y = check_X_y_pd_np_mismatch(X, y)
else:
   X = is_dataframe(X)
```

This means, taking the _is_dataframe out of _check_input_and_variables() hidden method in the base_encoder and amending the code slightly in all the classes.